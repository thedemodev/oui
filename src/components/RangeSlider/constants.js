import {
  amberBase,
  aquaBase,
  brandBlueDark,
  greenBase,
  orangeBase,
  pinkBase,
  redBase,
  magentaBase,
  greyBase,
} from '../../tokens/forimport/index.es';

// map fillColorName prop values to OUI color tokens
export const FILL_COLOR_MAP = {
  aqua: aquaBase,
  amber: amberBase,
  default: brandBlueDark,
  green: greenBase,
  orange: orangeBase,
  pink: pinkBase,
  red: redBase,
  magenta: magentaBase,
  grey: greyBase,
};

export default {
  FILL_COLOR_MAP,
};
