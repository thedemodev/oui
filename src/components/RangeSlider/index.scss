////
/// RangeSlider
/// @description Visual range selector
///
/// @example[html] Visual range slider
/// <div class="oui-rangeslider">
///   <div class="lego-grid">
///     <div class="lego-grid__cell position--relative">
///       <div class="flex">
///         <label class="lego-label muted flex--1">0%</label>
///         <label class="lego-label muted">100%</label>
///       </div>
///       <div class="range-display">
///         <div class="range-display range-display-overlay"></div>
///       </div>
///       <div class="range-grid">
///         <div class="gridline grid-1"></div>
///         <div class="gridline grid-2"></div>
///         <div class="gridline grid-3"></div>
///         <div class="gridline grid-4"></div>
///         <div class="gridline grid-5"></div>
///         <div class="gridline gridline-half grid-h1"></div>
///         <div class="gridline gridline-half grid-h2"></div>
///         <div class="gridline gridline-half grid-h3"></div>
///         <div class="gridline gridline-half grid-h4"></div>
///       </div>
///       <input class="hard position--relative width--1-1" data-traffic-group="1" type="range" min="0" max="100" step="1" value="50" oninput="resizeRangeDisplay(this.value)" />
///     </div>
///   </div>
/// </div>

.oui-rangeslider {
  input[type="range"] {
    -webkit-appearance: none;
    height: 50px;
    background: transparent;
  }
  input[type="range"]:focus {
    outline: none;
  }
  input[type="range"]::-webkit-slider-runnable-track {
    position: relative;
    top: 16px;
  }
  input[type="range"]::-webkit-slider-thumb {
    -webkit-appearance: none;
    border: 4px solid white;
    cursor: pointer;
    margin-top: -0.3em;
    width: 24px;
    height: 24px;
    background: white;
    background-size: 100%;
    background-image: linear-gradient(rgba(0, 0, 0, 0.13), rgba(255, 255, 255, 0));
    box-shadow: 0 0 4px rgba(110, 110, 110, 0.8), 0 0 2px rgba(0, 0, 0, 0.3);
    border-radius: 50%;
  }
  input[type="range"]:focus::-webkit-slider-thumb {
    box-shadow: 0 0 8px rgba(0, 116, 180,.9);
  }
  .range-display {
    height: 35px;
    position: absolute;
    left: 12px;
    right: 11px;
  }
  .range-display svg {
    display: inline-flex;
    width: 100%;
    height: 100%;
  }
  &.oui-rangeslider--disabled {
    .range-display {
      filter: grayscale(1) opacity(0.3);
    }
    .range-grid .gridline, .range-labels {
      opacity: 0.5;
    }
  }
  .range-display.range-display-overlay {
    background-color: white;
    background-image: none;
    display: inline-flex;
    left: 50%;
    opacity: 0.9;
    right: 0;
  }
  .range-display.range-display-overlay::after {
    display: none;
  }
  .range-display::after {
    content: '';
    position: absolute;
    left: 0;
    right: 0px;
    height: 8px;
    background-color: map-fetch($color, background, faint);
    border: 1px solid map-fetch($button, type, base, border);
    bottom: -8px;
    border-bottom-left-radius: 100px;
    border-bottom-right-radius: 100px;
    border-top: none;
    box-shadow: 1px 1px 10px rgba(0, 0, 0, 0.2) inset;
  }
  .range-grid {
    position: absolute;
    width: calc(100% - 24px);
    margin-left: 12px;
    height: 38px;
    top: 20px;
  }
  .range-grid .gridline {
    background-color: map-fetch($button, type, base, border);
    opacity: 1;
    position: absolute;
    left: 0px;
    right: 0px;
    width: 1px;
    height: 95%;
    bottom: 0px;
  }
  .range-grid .gridline.gridline-half {
    left: 0px;
    right: 0px;
    width: 1px;
    height: 50%;
  }
  .range-grid .gridline.grid-2 {
    left: 25%;
  }
  .range-grid .gridline.grid-3 {
    left: 50%;
  }
  .range-grid .gridline.grid-4 {
    left: 75%;
  }
  .range-grid .gridline.grid-5 {
    left: 100%;
  }
  .range-grid .gridline.grid-h1 {
    left: 12.5%;
  }
  .range-grid .gridline.grid-h2 {
    left: 37.5%;
  }
  .range-grid .gridline.grid-h3 {
    left: 62.5%;
  }
  .range-grid .gridline.grid-h4 {
    left: 87.5%;
  }
}
