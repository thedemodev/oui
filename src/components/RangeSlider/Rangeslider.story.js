import React from 'react';

import { storiesOf } from '@storybook/react';
import { withKnobs, number, select } from '@storybook/addon-knobs';

import RangeSlider from './index.js';
import { FILL_COLOR_MAP } from './constants';

// build fillColorName options for a select/dropdown knob
const fillColorOptions = { None: '' };
Object.keys(FILL_COLOR_MAP).forEach(color => {
  fillColorOptions[color] = color;
});

const stories = storiesOf('RangeSlider', module);
stories
  .addDecorator(withKnobs)
  .addDecorator(story => (
    <div id="root-preview">
      {story()}
    </div>
  ));

stories.add('Default', (() => {
  return (
    <RangeSlider
      value={ number('value', 50) }
    />
  );
}))
  .add('Disabled', (() => {
    return (
      <div>
        <input type="text" className="oui-text-input push-double--bottom" />
        <RangeSlider
          isDisabled={ true }
          value={ number('value', 50) }
        />
        <input type="text" className="oui-text-input" />
      </div>
    );
  }))
  .add('Custom Fill Color', (() => {
    return (
      <div>
        <RangeSlider
          fillColorName={ select('fillColorName', fillColorOptions, 'red') }
          value={ number('value', 50) }
        />
      </div>
    );
  }));
