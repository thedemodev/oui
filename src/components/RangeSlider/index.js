import PropTypes from 'prop-types';
import React from 'react';
import classnames from 'classnames';

import { FILL_COLOR_MAP } from './constants';

const DEFAULT_FILL_COLOR_NAME = 'default';

const RangeSlider = (props) => {
  const {
    fillColorName,
    value,
    onChange,
    isDisabled,
  } = props;

  const rangeClasses = classnames({
    'oui-rangeslider': true,
    'oui-rangeslider--disabled': isDisabled,
  });

  // ensure valid fillColor name (in the case that propType errors are ignored)
  const validFillColorName = Object.keys(FILL_COLOR_MAP).includes(fillColorName) ? fillColorName : DEFAULT_FILL_COLOR_NAME;
  const fillColor = FILL_COLOR_MAP[validFillColorName];

  return (
    <div className={ rangeClasses } data-test-section={ props.testSection }>
      <div className="oui-grid">
        <div className="oui-grid__cell position--relative">
          <div className="range-labels flex">
            <label className="oui-label muted flush flex--1">0%</label>
            <label className="oui-label muted flush">100%</label>
          </div>
          <div className="range-display">
            <svg
              preserveAspectRatio="none"
              xmlns="http://www.w3.org/2000/svg"
              width="256"
              height="256"
              viewBox="0 0 256 256">
              <path fill={ fillColor } d="M0 256h256V0z"/>
            </svg>
            <div
              className="range-display range-display-overlay"
              style={{ left: value + '%' }}>
            </div>
          </div>
          <div className="range-grid">
            <div className="gridline grid-1"></div>
            <div className="gridline grid-2"></div>
            <div className="gridline grid-3"></div>
            <div className="gridline grid-4"></div>
            <div className="gridline grid-5"></div>
            <div className="gridline gridline-half grid-h1"></div>
            <div className="gridline gridline-half grid-h2"></div>
            <div className="gridline gridline-half grid-h3"></div>
            <div className="gridline gridline-half grid-h4"></div>
          </div>
          <input
            className="hard position--relative width--1-1"
            data-traffic-group="1"
            type="range"
            min="0"
            max="100"
            step="1"
            value={ value }
            onChange={ onChange }
            disabled={ isDisabled }
          />
        </div>
      </div>
    </div>
  );
};

RangeSlider.propTypes = {
  /** Custom color for slider **/
  fillColorName: PropTypes.oneOf(Object.keys(FILL_COLOR_MAP)),
  /** Whether it is disabled, will render as greyscale if so **/
  isDisabled: PropTypes.bool,
  /** onChange function */
  onChange: PropTypes.func,
  /** Hook for automated JavaScript tests */
  testSection: PropTypes.string,
  /** The value */
  value: PropTypes.number,
};

RangeSlider.defaultProps = {
  fillColorName: DEFAULT_FILL_COLOR_NAME,
  isDisabled: false,
  onChange: undefined,
  testSection: undefined,
  value: undefined,
};

export default RangeSlider;
