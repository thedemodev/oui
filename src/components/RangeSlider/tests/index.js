import React from 'react';
import RangeSlider from '../index';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import { redBase } from '../../../tokens/forimport/index.es';

describe('components/RangeSlider', () => {
  let component;
  let handlerObj;

  describe('default fill color', () => {
    beforeEach(() => {
      handlerObj = {
        onChange: () => {},
      };
      spyOn(handlerObj, 'onChange');
      component = shallow(
        <RangeSlider
          value={ 50 }
          onChange={ handlerObj.onChange }
          testSection="myTestSection"
        />
      );
    });

    it('should render component same as snapshot', () => {
      expect(shallowToJson(component)).toMatchSnapshot();
    });

    it('should render component with correct props', () => {
      expect(component.length).toBe(1);
    });

    it('should set the input value from the value prop', () => {
      const input = component.find('input');
      expect(input.prop('value')).toBe(50);
    });

    it('should set the position of the overlay from the value prop', () => {
      const overlay = component.find('.range-display-overlay');
      expect(overlay.prop('style').left).toBe('50%');
    });

    it('should call the onChange prop when the input onChange is called', () => {
      component.find('input').prop('onChange')();
      expect(handlerObj.onChange).toHaveBeenCalled();
    });

    it('should set a data-test-section attribute from the testSection prop', () => {
      const testSection = component.find('[data-test-section="myTestSection"]');
      expect(testSection.length).toBe(1);
    });
  });

  describe('custom fill color', () => {
    beforeEach(() => {
      handlerObj = {
        onChange: () => {},
      };
      spyOn(handlerObj, 'onChange');
      component = shallow(
        <RangeSlider
          fillColorName="red"
          value={ 50 }
          onChange={ handlerObj.onChange }
          testSection="myTestSection"
        />
      );
    });

    it('should render the SVG with the fill color that corresponds to the provided fill color name', () => {
      const svgPath = component.find('path');
      expect(svgPath.prop('fill')).toBe(redBase);
    });
  });
});
