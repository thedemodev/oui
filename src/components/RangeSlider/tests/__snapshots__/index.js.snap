// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/RangeSlider default fill color should render component same as snapshot 1`] = `
<div
  className="oui-rangeslider"
  data-test-section="myTestSection"
>
  <div
    className="oui-grid"
  >
    <div
      className="oui-grid__cell position--relative"
    >
      <div
        className="range-labels flex"
      >
        <label
          className="oui-label muted flush flex--1"
        >
          0%
        </label>
        <label
          className="oui-label muted flush"
        >
          100%
        </label>
      </div>
      <div
        className="range-display"
      >
        <svg
          height="256"
          preserveAspectRatio="none"
          viewBox="0 0 256 256"
          width="256"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 256h256V0z"
            fill="hsla(227, 100%, 50%, 1)"
          />
        </svg>
        <div
          className="range-display range-display-overlay"
          style={
            Object {
              "left": "50%",
            }
          }
        />
      </div>
      <div
        className="range-grid"
      >
        <div
          className="gridline grid-1"
        />
        <div
          className="gridline grid-2"
        />
        <div
          className="gridline grid-3"
        />
        <div
          className="gridline grid-4"
        />
        <div
          className="gridline grid-5"
        />
        <div
          className="gridline gridline-half grid-h1"
        />
        <div
          className="gridline gridline-half grid-h2"
        />
        <div
          className="gridline gridline-half grid-h3"
        />
        <div
          className="gridline gridline-half grid-h4"
        />
      </div>
      <input
        className="hard position--relative width--1-1"
        data-traffic-group="1"
        disabled={false}
        max="100"
        min="0"
        onChange={[Function]}
        step="1"
        type="range"
        value={50}
      />
    </div>
  </div>
</div>
`;
